cmd/ndm_daemonset/probe/eventhandler.go

```diff
@@ -110,3 +153,254 @@ func (pe *ProbeEvent) initOrErrorEvent() {
+		2. A new disk is added to this node -  the disk was already present in the cluster and it was moved to this node
+		3. A disk was detached and reconnected to this node
+		4. An add event due to partition table reread . This may cause events to be generated without the disk
+			being physically removed this node - (when a new partition is created on the device also, its the same case)
+	*/
+
+	// check if the disk can be uniquely identified. we try to generate the UUID for the device
+	klog.V(4).Infof("checking if device: %s can be uniquely identified", bd.DevPath)
+	uuid, ok := generateUUID(bd)
+	if ok {
+		bd.UUID = uuid
+		klog.V(4).Infof("uuid: %s has been generated for device: %s", uuid, bd.DevPath)
+		bdAPI, err := pe.Controller.GetBlockDevice(uuid)
+
+		if errors.IsNotFound(err) {
+			klog.V(4).Infof("device: %s, uuid: %s not found in etcd", bd.DevPath, uuid)
+			/*
+				Cases when the BlockDevice is not found in etcd
+				1. The device is appearing in this cluster for the first time
+				2. The device had partitions and BlockDevice was not created
+			*/
+
+			if bd.DeviceAttributes.DeviceType == libudevwrapper.UDEV_PARTITION {
+				klog.V(4).Infof("device: %s is partition", bd.DevPath)
+				klog.V(4).Info("checking if device has a parent")
+				// check if device has a parent that is claimed
+				parentBD, ok := pe.Controller.BDHierarchy[bd.DependentDevices.Parent]
+				if !ok {
+					klog.V(4).Infof("unable to find parent device for device: %s", bd.DevPath)
+					return fmt.Errorf("cannot get parent device for device: %s", bd.DevPath)
+				}
+
+				klog.V(4).Infof("parent device: %s found for device: %s", parentBD.DevPath, bd.DevPath)
+				klog.V(4).Infof("checking if parent device can be uniquely identified")
+				parentUUID, parentOK := generateUUID(parentBD)
+				if !parentOK {
+					klog.V(4).Infof("unable to generate UUID for parent device, may be a device without WWN")
+					// cannot generate UUID for parent, may be a device without WWN
+					// used the new algorithm to create partitions
+					if len(bd.DependentDevices.Holders) > 0 {
+						klog.V(4).Infof("device: %s has holder devices %+v", bd.DevPath, bd.DependentDevices.Holders)
+						klog.V(4).Infof("skip creating BlockDevice resource")
+						return nil
+					}
+					// create BlockDevice resource for partition
+					klog.V(4).Infof("creating BlockDevice resource for device: %s with uuid: %s", bd.DevPath, bd.UUID)
+					deviceInfo := pe.Controller.NewDeviceInfoFromBlockDevice(&bd)
+					existingBlockDeviceResource := pe.Controller.GetExistingBlockDeviceResource(bdAPIList, deviceInfo.UUID)
+					err := pe.Controller.PushBlockDeviceResource(existingBlockDeviceResource, deviceInfo)
+					if err != nil {
+						klog.Error(err)
+						return err
+					}
+					return nil
+				}
+
+				klog.V(4).Infof("uuid: %s generated for parent device: %s", parentUUID, parentBD.DevPath)
+
+				parentBDAPI, err := pe.Controller.GetBlockDevice(parentUUID)
+
+				if errors.IsNotFound(err) {
```

Comment:
Is this block of the code repeating for cases when generate uuid is not found or if the parent doesn't have a block device.

Reply:
Yes @kmova . The code block is same. Just that log messages are different. Will see if I can combine them.

Reply:
Done.